### DISCUSSION TITLE
Slow operations in degraded cluster

### DISCUSSION THREAD

**Discussion by kepakiano** [rabbitmq/rabbitmq-server#10100] [Other] [answered]
### Describe the bug

Hi everyone,

I'm following up on [this discussion](https://github.com/rabbitmq/rabbitmq-server/discussions/9445). We have looked into it a little further and can reliably reproduce the performance issues we run into in case one node of a three node cluster fails.

### Reproduction steps

1. Create cluster consisting of three nodes (RabbitMQ version 3.12.10)
2. Open a connection to node 1
3. Kill node 2
4. Declare a classic queue using the open connection

### Expected behavior

Usually, declaring a classic queue takes less than 50ms. We expect this to still be true after one of the nodes has unexpectedly crashed (or has been killed manually).

### Additional context

We have created a Java integration test using the testcontainers framework. In the test, three containers sharing the same Erlang cookie are joined into a cluster. We then execute `rabbitmqctl cluster_status` which takes around five seconds. Afterwards, we declare a classic queue, which takes a couple dozen milliseconds as expected. So the cluster works and fulfills our requirements.
Now we introduce the disruption: We execute `docker kill` to kill one of the two containers we do not have an open AMQP connection to.
Then we again ask for the cluster status, which takes over 3 minutes and declare a classic queue which takes more than 8 seconds.
We assume that the cluster tries to reach the failed node during both these operations, runs into a timeout and only then reports success. Maybe there is a way to not ask the failed node?
We can manually achieve this by evicting the failed node from the cluster. After that, all operations are fast again.

I attached the log of our test. The interesting lines are:
line 16: this is how long the initial `rabbitmqctl cluster_status` took in the healthy cluster and below is the output
line 98: declaring the queue in the healthy cluster is fast
line 106: `rabbitmqctl cluster_status` after killing one node takes four minutes
line 181: declaring a queue after killing one node takes 8 seconds
line 185: evicting the failed node
line 192: `rabbitmqctl cluster_status`  in the 2 node cluster is fast again
line 265: declaring a queue is fast again (23ms)

[rabbitmq_log.txt](https://github.com/rabbitmq/rabbitmq-server/files/13647111/rabbitmq_log.txt)

**Comment by michaelklishin** [comment, marked as answer]
@kepakiano there isn't much that can be added to https://github.com/rabbitmq/rabbitmq-server/discussions/9445 and https://github.com/rabbitmq/rabbitmq-server/discussions/9522.

It's great that you expect operations in a cluster with down nodes to work exactly like they do in a cluster with every member up. But that's not how things work in practice.

[Peer unavailability takes time to detect](https://rabbitmq.com/nettick.html), this is true for every part of RabbitMQ. The only way to avoid that is to avoid features or configurations where nodes have to contact their peers. Such as a node-local queue replica placement strategy. This was recommended to you in https://github.com/rabbitmq/rabbitmq-server/discussions/944 but you seemingly want someone to wave a magic wand.

https://github.com/rabbitmq/rabbitmq-server/pull/9874 or even https://github.com/rabbitmq/rabbitmq-server/pull/10065 do not fundamentally change any of that, only chip away at the amount of time taken by any operations that may have to interact with a down node.

**Comment by kepakiano** [comment]
@michaelklishin Thank you for your answer! I hope you did not mistake me saying "I expect" as a demand, it definitely isn't. 

I would, however, appreciate some help clearing up some things I apparently don't understand yet. For example, how #944 is related here. It is my understanding that declaring a default classic queue (exclusive, auto-delete, non-durable) is by default something local to the node I'm connected to. Is that correct? If not (i.e. it's a cluster-wide operation), is there a configurable timeout which contributes to the delay of 8 seconds I'm seeing? (Additional info: I wanted to exclude the client as a source of error, for example if it called `GET /queues` when declaring, so I used `rabbitmqadmin` to directly declare the queue on a node and the delay persists). In #9445 you suggested we gather information on when this issue occurs and also said that delays of >5s could be problematic, which is why we opened this issue/discussion.

For now, we can ignore the delay seen when querying the cluster status. It's a curiosity, but wouldn't break our use case.

**Reply by kjnilsson** [reply]
@kepakiano you didn't mention earlier that the queues were exclusive queues and you are right they are always local to the connection and thus should not need to run leader location logic (which may query other nodes). I will take a look at that code path.

Could you share the output of a cluster_status command that took a long time? I think the one you shared above was a "fast" run...

**Reply by kepakiano** [reply]
Ah, sorry, should have mentioned it was a default `queueDeclare` earlier. You should already find the output you're looking for in the log: The output of the fast cluster_status starts on line 18 and the slow one on line 107 (line 106 says `cluster_status` took >4min).

**Reply by kjnilsson** [reply]
so we have found a bit of code that very well could be responsible for the 8s delay when declaring exclusive queues: 

https://github.com/rabbitmq/rabbitmq-server/pull/10102

Also I found some improvements to the `cluster_status` command which may help a bit but 4 minutes sounds a lot longer than the 1 minute we'd expect it to take to detect the dangling TCP connection.

https://github.com/rabbitmq/rabbitmq-server/pull/10101

Do you use TLS between nodes? Can you tell us anything about your networking that may potentially make TCP connection setup and/or DNS resolution take a long time?

**Reply by kjnilsson** [reply]
Ok I have now managed to a local reproduction of what I think the issue with cluster_status is and I am pretty confident I know what needs to change.

**Reply by stefanlin** [reply]
@kjnilsson maybe you can give a hint in which part in the code path you expecting the issue. I try to familiarize myself with the topic a little.

**Reply by michaelklishin** [reply]
We have found a few non-obvious calls to other nodes in places that do not really need them. A PR or two will materialize later this week.

**Reply by kepakiano** [reply]
@kjnilsson We do not use TLS between the nodes, neither in test nor in prod. DNS resolution should not be a problem as it's provided by K8s in prod and in the test it was on the same host. This means that the TCP connections should establish quickly in the test and fairly quickly in prod: The nodes run in different datacenters with a inter-datacenter latency of <1 ms

**Comment by kjnilsson** [comment]
Ok so the cluster_status slowness should be mostly addressed in https://github.com/rabbitmq/rabbitmq-server/pull/10101

The slowness of classic queue declarations should be addressed in https://github.com/rabbitmq/rabbitmq-server/pull/10102

Testing welcome @kepakiano !

**Reply by kepakiano** [reply]
Awesome, thank you @kjnilsson ! Does the github pipeline already put a Docker image for this branch somewhere or should I build this myself?

**Reply by kjnilsson** [reply]
it does yes, let me find some links.

**Reply by mkuratczyk** [reply]
https://hub.docker.com/r/pivotalrabbitmq/rabbitmq/tags

and search for the branch name

**Reply by kjnilsson** [reply]
cluster status: https://hub.docker.com/layers/pivotalrabbitmq/rabbitmq/cluster_status_opt-otp-max-bazel/images/sha256-a77bc161787539e533c36a2c9aa6700952fdbcd283a3df84ee9872baca324873?context=explore

and

queue declare: https://hub.docker.com/layers/pivotalrabbitmq/rabbitmq/mirrored-otp-max-bazel/images/sha256-f51064ec7f634966655d0203bd94e805a73fc66b7401e343d929a3d4a88c98c9?context=explore

**Reply by kjnilsson** [reply]
So some expectations: `cluster_status` will still be a bit slow, on my local testing I went from 3 minutes to ~20s. If you have many feature flags that aren't enabled it may be slower. 

Queue declarations should be reasonably fast though. If not let us know.

**Reply by kepakiano** [reply]
Thank you for providing the images! Nice pipeline.
However, we haven't been able to see the improvement. We first used `pivotalrabbitmq/rabbitmq:mirrored-otp-max-bazel` in our Java testcontainers test and saw that declaring the queue was slow after one node had failed. We weren't 100% sure we were using the correct image (just double checking), so we're now trying to create the cluster manually, which does not work. Since `mirrored` has been merged into `main`, we start three nodes from `pivotalrabbitmq/rabbitmq:main-otp-max-bazel`:
```
CONTAINER ID        IMAGE                                         COMMAND                  CREATED             STATUS              PORTS                                                                                                                      NAMES
f2638c28185b        pivotalrabbitmq/rabbitmq:main-otp-max-bazel   "docker-entrypoint.s…"   5 minutes ago       Up 5 minutes        1883/tcp, 4369/tcp, 5551-5552/tcp, 5671-5672/tcp, 8883/tcp, 15670-15676/tcp, 15691-15692/tcp, 25672/tcp, 61613-61614/tcp   rabbit3
b73541d18f7f        pivotalrabbitmq/rabbitmq:main-otp-max-bazel   "docker-entrypoint.s…"   5 minutes ago       Up 5 minutes        1883/tcp, 4369/tcp, 5551-5552/tcp, 5671-5672/tcp, 8883/tcp, 15670-15676/tcp, 15691-15692/tcp, 25672/tcp, 61613-61614/tcp   rabbit2
99cdecc87c81        pivotalrabbitmq/rabbitmq:main-otp-max-bazel   "docker-entrypoint.s…"   6 minutes ago       Up 6 minutes        1883/tcp, 4369/tcp, 5551-5552/tcp, 5671-5672/tcp, 8883/tcp, 15670-15676/tcp, 15691-15692/tcp, 25672/tcp, 61613-61614/tcp   rabbit1
```

and then log into rabbit2 to join the rabbit1's cluster, but this crashes:
```
docker exec -it b7 /bin/bash
root@b73541d18f7f:/# rabbitmqctl stop_app && rabbitmqctl join_cluster 99cdecc87c81 && rabbitmqctl start_app
Stopping rabbit application on node rabbit@b73541d18f7f ...
Clustering node rabbit@b73541d18f7f with 99cdecc87c81

12:44:02.678 [error] Feature flags: error while running:
Feature flags:   rabbit_ff_controller:running_nodes[]
Feature flags: on node `99cdecc87c81@b73541d18f7f`:
Feature flags:   exception error: {erpc,noconnection}
Feature flags:     in function  erpc:call/5 (erpc.erl, line 710)
Feature flags:     in call from rabbit_ff_controller:rpc_call/5 (rabbit_ff_controller.erl, line 1377)
Feature flags:     in call from rabbit_ff_controller:list_nodes_clustered_with/1 (rabbit_ff_controller.erl, line 477)
Feature flags:     in call from rabbit_ff_controller:check_node_compatibility_task/2 (rabbit_ff_controller.erl, line 389)
Feature flags:     in call from rabbit_db_cluster:can_join/1 (rabbit_db_cluster.erl, line 65)
Feature flags:     in call from rabbit_db_cluster:join/2 (rabbit_db_cluster.erl, line 97)
Feature flags:     in call from erpc:execute_call/4 (erpc.erl, line 589)

Error:
{:aborted_feature_flags_compat_check, {:error, {:erpc, :noconnection}}}
```

Have you seen this error before?


Sporadically we also see this error when stopping the rabbit app:
```
Monotonic time stepped backwards!
Previous time: 1453237446
Current time:  1453185295
Aborted (core dumped)
```
In all test cases all three containers are started on the same host.

**Reply by michaelklishin** [reply]
https://github.com/rabbitmq/rabbitmq-server/pull/10102 has optimized classic (mirrored) queue declaration, not deletion.

[`rabbit_ff_controller`](https://github.com/rabbitmq/rabbitmq-server/commits/main/deps/rabbit/src/rabbit_ff_controller.erl) hasn't changed in two months.
It could handle this specific `erpc:call/5` exception but that's not related to the topic of this discussion.

**Reply by kepakiano** [reply]
Sorry, declaring, of course. Fixed it in the previous comment.

**Reply by kjnilsson** [reply]
@kepakiano it may well be that it is mnesia that is slowing down operations, once you've got your cluster sorted if you try enabling the `khepri_db` feature flag (i.e. moving to the new khepri / raft based meta data store) the you may get a much better reponsiveness in this scenario when it comes to any action that write to the meta data store (create/delete queue, exchange, binding, user, etc).

**Reply by kjnilsson** [reply]
monotonic time going backwards is a very bad thing and is something that happens if the operating systems monotonic time goes backwards (should not happen). The erlang VM cannot handle this so exits. Perhaps there are time sync issues?

**Reply by dumbbell** [reply]
@kepakiano: Note that you don't need to run `stop_app`/`start_app` around a `join_cluster` anymore when using the `main` branch. RabbitMQ will take care of stopping itself if needed.

**Reply by kepakiano** [reply]
Oh sweet, thank you @dumbbell ! Have a meeting now, will get back to you asap.

**Reply by michaelklishin** [reply]
The `erpc` exception was observed before https://github.com/rabbitmq/rabbitmq-server/issues/8346 without a clear understanding what specifically triggers it or how this behavior can be reproduced reasonably reliably.

**Reply by kepakiano** [reply]
@michaelklishin I can tell you: The reasons is users not using the CLI tools correctly. I forgot to put "rabbit@" in front of the name of the node I was joining...

**Reply by michaelklishin** [reply]
@kepakiano so a sequence of the `stop_app`, `join_cluster`, `start_app` with an incorrect node name with some commands reliably triggers it for you?

I can see a connection there, every CLI command opens a connection to a node and it is very short lived. The feature flag module may consider all connected nodes, including the CLI tool connection.

**Reply by michaelklishin** [reply]
At the same time, if the node name is completely off, there won't even be a connection.

Anyhow, we would really appreciate a step-by-step set of instructions that reproduces the exception some of the time as a comment in #8346.

**Reply by kepakiano** [reply]
yes, correct, I could've seen it before in the log I posted earlier:
```
12:44:02.678 [error] Feature flags: error while running:
Feature flags:   rabbit_ff_controller:running_nodes[]
Feature flags: on node `99cdecc87c81@b73541d18f7f`:
```


On the other matter: I've been testing on the main branch now after both fixes have been merged. Unfortunately, the fixes do not seem to help in every case: Sometimes when I execute the test it's blazingly fast and sometimes it is as before. Hopefully, I'll be able to provide some logs tomorrow. It looks as if in case of a slow run, the failed node is still part of the "disk node" section in the cluster_status, but that might just be spurious correlation. Furthermore, we've noticed that sometimes creating a channel in the failed state takes more than 3 seconds. Again, more logs on that tomorrow.

**Reply by kepakiano** [reply]
@michaelklishin will provide that along with our code test code tomorrow :)

**Reply by kepakiano** [reply]
Alright, so we did observe that declaring the queue was not slow every time we started our test (which involved starting a cluster, killing a node and then declaring a queue). We then moved from declaring one queue in a test to declaring 1000 and recording the times it took to declare a queue. As you can see in the log, out of 1000 declarations 4 took longer than 1s and they were all pretty close to 8 seconds. If the declaration took less than 1 second the average is 23ms with 4ms min and 254ms max.
Obviously, this is a great improvement over the old version (if you use the old image, 100% of the declarations take longer than 1s and the average is 15s) and 0.4% of the declarations being slow might already be good enough for our use case.

This is the log: [test_output.txt](https://github.com/rabbitmq/rabbitmq-server/files/13674766/test_output.txt)

You can find the code for our test [here](https://github.com/anjoTorrado/rabbitMqTest)
A few notes: 
- `rabbitMq2.stop()` does actually kill the container despite its name
- Testcontainers has some restrictions on the image name so we need to re-tag it manually, see the comment in the code
- Sometimes, creating a channel in a cluster with a failed node using a connection that had been opened before killing the node takes a significant amount of time (we observed 3.5s), but this does not correlate with slow queue declaration
- If declaring a queue is slow, getting the cluster_status is also slow! ~3-5 minutes

**Reply by kjnilsson** [reply]
@kepakiano are you testing off an image from the `main` branch?

**Reply by kepakiano** [reply]
Yes, we're testing on `main-otp-max-bazel` from today

**Reply by kjnilsson** [reply]
odd, cluster_status shouldn't be that slow in that case with the PR being merged yesterday.

**Reply by kepakiano** [reply]
Indeed. Are you able to reproduce the problem with the code above?

**Reply by kjnilsson** [reply]
```17:12:23.320 tc.testcontainers/ryuk:0.5.1 INFO   Creating container for image: testcontainers/ryuk:0.5.1
17:12:23.323 org.testcontainers.utility.RegistryAuthLocator INFO   Failure when attempting to lookup auth config. Please ignore if you don't have images in an authenticated registry. Details: (dockerImageName: testcontainers/ryuk:0.5.1, configFile: /home/karl/.docker/config.json, configEnv: DOCKER_AUTH_CONFIG). Falling back to docker-java default behaviour. Exception message: Status 404: No config supplied. Checked in order: /home/karl/.docker/config.json (file not found), DOCKER_AUTH_CONFIG (not set)
17:12:23.559 tc.testcontainers/ryuk:0.5.1 INFO   Container testcontainers/ryuk:0.5.1 is starting: ce9007bfe3462f221a4231eb228f11bfaf2574281b0cbc90fe9867b4d853107c
17:12:24.074 tc.testcontainers/ryuk:0.5.1 INFO   Container testcontainers/ryuk:0.5.1 started in PT0.753844645S
17:12:24.078 org.testcontainers.utility.RyukResourceReaper INFO   Ryuk started - will monitor and terminate Testcontainers containers on JVM exit
17:12:24.079 org.testcontainers.DockerClientFactory INFO   Checking the system...
17:12:24.079 org.testcontainers.DockerClientFactory INFO   ✔︎ Docker server version should be at least 1.6.0
17:12:24.080 tc.rabbitmq:main-otp-max-bazel INFO   Creating container for image: rabbitmq:main-otp-max-bazel
17:12:24.230 tc.rabbitmq:main-otp-max-bazel INFO   Container rabbitmq:main-otp-max-bazel is starting: 18d4d50fdfc3fd9c8aa7641227a03e374b3fd6a6262f65c324ca1706364ee016
17:12:29.712 tc.rabbitmq:main-otp-max-bazel INFO   Container rabbitmq:main-otp-max-bazel started in PT5.632102245S
17:12:29.712 RabbitMqClusterTest INFO   rabbitmq1 started
17:12:29.713 tc.rabbitmq:main-otp-max-bazel INFO   Creating container for image: rabbitmq:main-otp-max-bazel
17:12:29.784 tc.rabbitmq:main-otp-max-bazel INFO   Container rabbitmq:main-otp-max-bazel is starting: 3b9b3b892aa6b2d23ea321db1822c1e6783e3d33498686eb68deec49eeb87ca9
17:12:35.170 tc.rabbitmq:main-otp-max-bazel INFO   Container rabbitmq:main-otp-max-bazel started in PT5.457256819S
17:12:35.170 RabbitMqClusterTest INFO   rabbitmq2 started
17:12:35.170 tc.rabbitmq:main-otp-max-bazel INFO   Creating container for image: rabbitmq:main-otp-max-bazel
17:12:35.235 tc.rabbitmq:main-otp-max-bazel INFO   Container rabbitmq:main-otp-max-bazel is starting: 3fd07695ecee7e30b282ba10bef0cbfe261972d7e7f968840761b1ac5cc6c28c
17:12:40.681 tc.rabbitmq:main-otp-max-bazel INFO   Container rabbitmq:main-otp-max-bazel started in PT5.510894927S
17:12:40.681 RabbitMqClusterTest INFO   rabbitmq3 started
17:12:40.693 RabbitMqClusterTest INFO   Setting up cluster
17:12:58.250 RabbitMqClusterTest INFO   Rabbitmq available on port 32770
17:12:58.265 RabbitMqClusterTest INFO   Killing: 3b9b3b892aa6
17:12:59.001 RabbitMqClusterTest INFO   Sleeping for 5s
17:13:04.001 RabbitMqClusterTest INFO   Killed 3b9b3b892aa6
17:13:04.002 RabbitMqClusterTest INFO   Cluster status after node2 has been killed:
17:13:04.109 RabbitMqClusterTest INFO   Declaring queue #0
17:13:04.112 RabbitMqClusterTest INFO   Declaring queue #1
17:13:04.114 RabbitMqClusterTest INFO   Declaring queue #2
17:13:04.116 RabbitMqClusterTest INFO   Declaring queue #3
17:13:04.117 RabbitMqClusterTest INFO   Declaring queue #4
17:13:04.119 RabbitMqClusterTest INFO   Declaring queue #5
17:13:04.120 RabbitMqClusterTest INFO   Declaring queue #6
17:13:04.122 RabbitMqClusterTest INFO   Declaring queue #7
17:13:04.124 RabbitMqClusterTest INFO   Declaring queue #8
17:13:04.125 RabbitMqClusterTest INFO   Declaring queue #9
17:13:04.127 RabbitMqClusterTest INFO   Declaring queue #10
17:13:04.128 RabbitMqClusterTest INFO   Declaring queue #11
17:13:04.130 RabbitMqClusterTest INFO   Declaring queue #12
17:13:04.132 RabbitMqClusterTest INFO   Declaring queue #13
17:13:04.134 RabbitMqClusterTest INFO   Declaring queue #14
17:13:04.135 RabbitMqClusterTest INFO   Declaring queue #15
17:13:04.137 RabbitMqClusterTest INFO   Declaring queue #16
17:13:04.139 RabbitMqClusterTest INFO   Declaring queue #17
17:13:04.140 RabbitMqClusterTest INFO   Declaring queue #18
17:13:04.142 RabbitMqClusterTest INFO   Declaring queue #19
17:13:04.143 RabbitMqClusterTest INFO   Declaring queue #20
17:13:04.145 RabbitMqClusterTest INFO   Declaring queue #21
17:13:04.146 RabbitMqClusterTest INFO   Declaring queue #22
17:13:04.148 RabbitMqClusterTest INFO   Declaring queue #23
17:13:04.149 RabbitMqClusterTest INFO   Declaring queue #24
17:13:04.152 RabbitMqClusterTest INFO   Declaring queue #25
17:13:04.153 RabbitMqClusterTest INFO   Declaring queue #26
17:13:04.155 RabbitMqClusterTest INFO   Declaring queue #27
17:13:04.156 RabbitMqClusterTest INFO   Declaring queue #28
17:13:04.158 RabbitMqClusterTest INFO   Declaring queue #29
17:13:04.159 RabbitMqClusterTest INFO   Declaring queue #30
17:13:04.161 RabbitMqClusterTest INFO   Declaring queue #31
17:13:04.163 RabbitMqClusterTest INFO   Declaring queue #32
17:13:04.164 RabbitMqClusterTest INFO   Declaring queue #33
17:13:04.166 RabbitMqClusterTest INFO   Declaring queue #34
17:13:04.167 RabbitMqClusterTest INFO   Declaring queue #35
17:13:04.168 RabbitMqClusterTest INFO   Declaring queue #36
17:13:04.170 RabbitMqClusterTest INFO   Declaring queue #37
17:13:04.172 RabbitMqClusterTest INFO   Declaring queue #38
17:13:04.174 RabbitMqClusterTest INFO   Declaring queue #39
17:13:04.176 RabbitMqClusterTest INFO   Declaring queue #40
17:13:04.178 RabbitMqClusterTest INFO   Declaring queue #41
17:13:04.179 RabbitMqClusterTest INFO   Declaring queue #42
17:13:04.181 RabbitMqClusterTest INFO   Declaring queue #43
17:13:04.182 RabbitMqClusterTest INFO   Declaring queue #44
17:13:04.183 RabbitMqClusterTest INFO   Declaring queue #45
17:13:04.185 RabbitMqClusterTest INFO   Declaring queue #46
17:13:04.186 RabbitMqClusterTest INFO   Declaring queue #47
17:13:04.187 RabbitMqClusterTest INFO   Declaring queue #48
17:13:04.189 RabbitMqClusterTest INFO   Declaring queue #49
17:13:04.192 RabbitMqClusterTest INFO   Declaring queue #50
17:13:04.194 RabbitMqClusterTest INFO   Declaring queue #51
17:13:04.196 RabbitMqClusterTest INFO   Declaring queue #52
17:13:04.198 RabbitMqClusterTest INFO   Declaring queue #53
17:13:04.200 RabbitMqClusterTest INFO   Declaring queue #54
17:13:04.201 RabbitMqClusterTest INFO   Declaring queue #55
17:13:04.203 RabbitMqClusterTest INFO   Declaring queue #56
17:13:04.205 RabbitMqClusterTest INFO   Declaring queue #57
17:13:04.208 RabbitMqClusterTest INFO   Declaring queue #58
17:13:04.210 RabbitMqClusterTest INFO   Declaring queue #59
17:13:04.212 RabbitMqClusterTest INFO   Declaring queue #60
17:13:04.214 RabbitMqClusterTest INFO   Declaring queue #61
17:13:04.215 RabbitMqClusterTest INFO   Declaring queue #62
17:13:04.217 RabbitMqClusterTest INFO   Declaring queue #63
17:13:04.218 RabbitMqClusterTest INFO   Declaring queue #64
17:13:04.219 RabbitMqClusterTest INFO   Declaring queue #65
17:13:04.221 RabbitMqClusterTest INFO   Declaring queue #66
17:13:04.222 RabbitMqClusterTest INFO   Declaring queue #67
17:13:04.223 RabbitMqClusterTest INFO   Declaring queue #68
17:13:04.225 RabbitMqClusterTest INFO   Declaring queue #69
17:13:04.226 RabbitMqClusterTest INFO   Declaring queue #70
17:13:04.227 RabbitMqClusterTest INFO   Declaring queue #71
17:13:04.229 RabbitMqClusterTest INFO   Declaring queue #72
17:13:04.230 RabbitMqClusterTest INFO   Declaring queue #73
17:13:04.232 RabbitMqClusterTest INFO   Declaring queue #74
17:13:04.233 RabbitMqClusterTest INFO   Declaring queue #75
17:13:04.235 RabbitMqClusterTest INFO   Declaring queue #76
17:13:04.236 RabbitMqClusterTest INFO   Declaring queue #77
17:13:04.238 RabbitMqClusterTest INFO   Declaring queue #78
17:13:04.239 RabbitMqClusterTest INFO   Declaring queue #79
17:13:04.241 RabbitMqClusterTest INFO   Declaring queue #80
17:13:04.242 RabbitMqClusterTest INFO   Declaring queue #81
17:13:04.244 RabbitMqClusterTest INFO   Declaring queue #82
17:13:04.245 RabbitMqClusterTest INFO   Declaring queue #83
17:13:04.246 RabbitMqClusterTest INFO   Declaring queue #84
17:13:04.248 RabbitMqClusterTest INFO   Declaring queue #85
17:13:04.249 RabbitMqClusterTest INFO   Declaring queue #86
17:13:04.250 RabbitMqClusterTest INFO   Declaring queue #87
17:13:04.252 RabbitMqClusterTest INFO   Declaring queue #88
17:13:04.253 RabbitMqClusterTest INFO   Declaring queue #89
17:13:04.254 RabbitMqClusterTest INFO   Declaring queue #90
17:13:04.256 RabbitMqClusterTest INFO   Declaring queue #91
17:13:04.257 RabbitMqClusterTest INFO   Declaring queue #92
17:13:04.259 RabbitMqClusterTest INFO   Declaring queue #93
17:13:04.260 RabbitMqClusterTest INFO   Declaring queue #94
17:13:04.262 RabbitMqClusterTest INFO   Declaring queue #95
17:13:04.263 RabbitMqClusterTest INFO   Declaring queue #96
17:13:04.264 RabbitMqClusterTest INFO   Declaring queue #97
17:13:04.266 RabbitMqClusterTest INFO   Declaring queue #98
17:13:04.267 RabbitMqClusterTest INFO   Declaring queue #99
17:13:04.268 RabbitMqClusterTest INFO   Declaring queue #100
17:13:04.270 RabbitMqClusterTest INFO   Declaring queue #101
17:13:04.271 RabbitMqClusterTest INFO   Declaring queue #102
17:13:04.272 RabbitMqClusterTest INFO   Declaring queue #103
17:13:04.274 RabbitMqClusterTest INFO   Declaring queue #104
17:13:04.275 RabbitMqClusterTest INFO   Declaring queue #105
17:13:04.277 RabbitMqClusterTest INFO   Declaring queue #106
17:13:04.278 RabbitMqClusterTest INFO   Declaring queue #107
17:13:04.279 RabbitMqClusterTest INFO   Declaring queue #108
17:13:04.281 RabbitMqClusterTest INFO   Declaring queue #109
17:13:04.282 RabbitMqClusterTest INFO   Declaring queue #110
17:13:04.285 RabbitMqClusterTest INFO   Declaring queue #111
17:13:04.287 RabbitMqClusterTest INFO   Declaring queue #112
17:13:04.288 RabbitMqClusterTest INFO   Declaring queue #113
17:13:04.291 RabbitMqClusterTest INFO   Declaring queue #114
17:13:04.295 RabbitMqClusterTest INFO   Declaring queue #115
17:13:04.297 RabbitMqClusterTest INFO   Declaring queue #116
17:13:04.299 RabbitMqClusterTest INFO   Declaring queue #117
17:13:04.301 RabbitMqClusterTest INFO   Declaring queue #118
17:13:04.302 RabbitMqClusterTest INFO   Declaring queue #119
17:13:04.305 RabbitMqClusterTest INFO   Declaring queue #120
17:13:04.307 RabbitMqClusterTest INFO   Declaring queue #121
17:13:04.309 RabbitMqClusterTest INFO   Declaring queue #122
17:13:04.312 RabbitMqClusterTest INFO   Declaring queue #123
17:13:04.313 RabbitMqClusterTest INFO   Declaring queue #124
17:13:04.315 RabbitMqClusterTest INFO   Declaring queue #125
17:13:04.316 RabbitMqClusterTest INFO   Declaring queue #126
17:13:04.317 RabbitMqClusterTest INFO   Declaring queue #127
17:13:04.319 RabbitMqClusterTest INFO   Declaring queue #128
17:13:04.320 RabbitMqClusterTest INFO   Declaring queue #129
17:13:04.321 RabbitMqClusterTest INFO   Declaring queue #130
17:13:04.322 RabbitMqClusterTest INFO   Declaring queue #131
17:13:04.324 RabbitMqClusterTest INFO   Declaring queue #132
17:13:04.325 RabbitMqClusterTest INFO   Declaring queue #133
17:13:04.327 RabbitMqClusterTest INFO   Declaring queue #134
17:13:04.329 RabbitMqClusterTest INFO   Declaring queue #135
17:13:04.330 RabbitMqClusterTest INFO   Declaring queue #136
17:13:04.332 RabbitMqClusterTest INFO   Declaring queue #137
17:13:04.333 RabbitMqClusterTest INFO   Declaring queue #138
17:13:04.334 RabbitMqClusterTest INFO   Declaring queue #139
17:13:04.336 RabbitMqClusterTest INFO   Declaring queue #140
17:13:04.338 RabbitMqClusterTest INFO   Declaring queue #141
17:13:04.339 RabbitMqClusterTest INFO   Declaring queue #142
17:13:04.341 RabbitMqClusterTest INFO   Declaring queue #143
17:13:04.342 RabbitMqClusterTest INFO   Declaring queue #144
17:13:04.344 RabbitMqClusterTest INFO   Declaring queue #145

17:13:05.650 RabbitMqClusterTest INFO   Declaring queue #995
17:13:05.651 RabbitMqClusterTest INFO   Declaring queue #996
17:13:05.652 RabbitMqClusterTest INFO   Declaring queue #997
17:13:05.653 RabbitMqClusterTest INFO   Declaring queue #998
17:13:05.655 RabbitMqClusterTest INFO   Declaring queue #999
17:13:05.657 RabbitMqClusterTest INFO   ##### > 1 second
17:13:05.657 RabbitMqClusterTest INFO   Amount of queues: 0
17:13:05.657 RabbitMqClusterTest INFO   Slowest queue declaration took 0 ms.
17:13:05.657 RabbitMqClusterTest INFO   Fastest queue declaration took 0 ms.
17:13:05.659 RabbitMqClusterTest INFO   Average queue declaration took 0.0 ms.
17:13:05.659 RabbitMqClusterTest INFO   ----------------------------------------------------------------
17:13:05.659 RabbitMqClusterTest INFO   ##### <= 1 second
17:13:05.659 RabbitMqClusterTest INFO   Amount of queues: 1000
17:13:05.660 RabbitMqClusterTest INFO   Slowest queue declaration took 6 ms.
17:13:05.660 RabbitMqClusterTest INFO   Fastest queue declaration took 0 ms.
17:13:05.661 RabbitMqClusterTest INFO   Average queue declaration took 1.166 ms.
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 43.50 s -- in RabbitMqClusterTest
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  44.920 s
[INFO] Finished at: 2023-12-14T17:13:06Z
[INFO] ------------------------------------------------------------------------
```

**Reply by kepakiano** [reply]
Hm... that is awkward. Maybe run it again or increase the amount of queues? 🙈

**Reply by michaelklishin** [reply]
This thread has quickly spiraled into troubleshooting everything everywhere all at once.

Using 20K queues will result in a slowdown over time for each subsequent declaration. Because existing queues will consume some resources, result in more frequent scheduler and CPU context switching, and so on.

It should not matter how many queues you have w.r.t. whether a non-replicated queue declaration contacts other nodes, unreachable nodes specifically.

The original complaint was that a few operations are slower than they probably could be when some cluster nodes are down. They have been optimized to the extent we can do quickly. We have provided evidence of that.

**Reply by michaelklishin** [reply]
There three improvements in total in the last month or so. The biggest gain will probably come from finishing https://github.com/rabbitmq/rabbitmq-server/pull/10065 (if and when that happens), then adopting it in more places.

This will not be a priority for our small team. We have customer escalations to assist to, 3.13 and 4.0 to ship, commercial features to work on. If you really need more improvements, you are welcome to contribute them. Especially given not three but four examples of how and in what areas we've done it previously.

**Reply by kepakiano** [reply]
Hey @michaelklishin! We are happy about the implemented fixes, as they obviously provide a big performance boost in the circumstances we demonstrated. We have provided evidence that there is still a chance for the performance of declaring a queue to be bad in some cases, but since the likelihood is so low that we will also not pursue this at the moment. 
Once #10065 has been implemented and v3.13 shipped we will run some more rounds of tests and possibly report back here.

**Reply by kjnilsson** [reply]
@kepakiano we appreciate your testing and efforts. I have run your test app several times on my machine and it is always fast. Could you have run against a stale set of docker images when you saw the `cluster_status` slowness?

**Reply by michaelklishin** [reply]
We offer no guarantees that #10065 will be implemented, or that it will ship in 3.13.0.

**Reply by kepakiano** [reply]
@kjnilsson I don't think it was an issue with a stale image, as we all removed the images we had and pulled them fresh last Wednesday. However, it could have been an issue with our test setups all being comprised of various virtualization methods, older Docker versions and not too powerful hardware. I just ran the test several times on a more powerful machine with native Linux and the issue does not seem to persist 🎉  So thank you for that :)

@michaelklishin no problem, we'll just keep our eyes open :)

**Reply by kjnilsson** [reply]
Ok thanks for reporting back. #10065 wont ship until 4.0 at the earliest as it does involve a certain degree of risk and 3.13 is about to go soon.

**Reply by kepakiano** [reply]
Alright, we're looking forward to it either way.
